pytoda/datasets/_table_dataset.py

```diff
@@ -152,5 +148,61 @@ def __getitem__(self, index: int) -> torch.tensor:
                 for the current sample.
         """
         return torch.tensor(
-            self._dataset[index], dtype=self.dtype, device=self.device
+            self.dataset[index], dtype=self.dtype, device=self.device
+        )  # base_dataset: TODO torch.tensor needed? Dataloader should take care of this.
+        # if not we can remove this
+
+
+class _TableLazyDataset(_TableDataset):
+    """
+    Table dataset using lazy loading.
+
+    Suggested when handling datasets that can fit in the device memory.
+    In case of datasets fitting in device memory consider using
+    _TableEagerDataset for better performance.
+    """
+
+    def _setup_dataset(self) -> None:
+        """Setup KeyDataset assigned to self.dataset for delegation."""
+        self.dataset = concatenate_file_based_datasets(
+            filepaths=self.filepaths,
+            dataset_class=_CsvLazyDataset,
+            feature_list=self.feature_list,
+            chunk_size=self.chunk_size,
+            **self.kwargs
         )
+
+    def _preprocess_dataset(self) -> None:
+        """Preprocess the dataset."""
+        self.feature_fn = lambda sample: sample[self.feature_mapping[
+            self.feature_list].values]
+        for dataset in self.dataset.datasets:
+            for index in dataset.cache:
+                dataset.cache[index] = self.transform_fn(
+                    self.feature_fn(dataset.cache[index])
+                )
+
+
+class _TableEagerDataset(_TableDataset):
+    """
+    Table dataset using eager loading.
+
+    Suggested when handling datasets that can fit in the device memory.
+    In case of out of memory errors consider using _TableLazyDataset.
+    """
+
+    def _setup_dataset(self) -> None:
+        """Setup KeyDataset assigned to self.dataset for delegation."""
+        self.dataset = concatenate_file_based_datasets(
+            filepaths=self.filepaths,
+            dataset_class=_CsvEagerDataset,
+            feature_list=self.feature_list,
+            dtype={'cell_line': str},
```

Comment:
I do not like this data-sepcific component in n otherwise generic class. Can we think of a nicer solution?. Maybe having this dtype controllable and defaulting to empty, and then having this cell_line in the datasets using/inheriting this class.

Reply:
Good catch. I don't think the argument is even needed/used by us, as it's not passed in the lazy version that reads the csv the same way but chunked.

But it can be quite relevant, so I added the option to pass `pandas_dtype` via kwargs (pretty much hidden) until it reaches _CsvStatistics, which put it as dtype into self.kwargs to be passed to pd.read_csv in both _CsvEagerDataset and __CsvLazyDataset.